add validation to LedgerIndex

tekbar11 opened the pull request:
Proposed fix for #93 .

Added a method `isValid` to LedgerIndex to check the validity of the wrapped index.

The `LedgerIndex.of` factory method now throws a `NumberFormatException` if the given value is not a valid index.

sappenin commented on xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java:
```diff
@@ -34,9 +34,17 @@ public LedgerIndex(String value) {
    */
-  public static LedgerIndex of(String value) {
-    return new LedgerIndex(value);
+  public static LedgerIndex of(String value)
```
Let's add some unit test coverage for the three text-values, and probably for a `null` value too?

tekbar11 replied:
I added some unit tests as requested.

I'd also like to point out that the `LedgeIndexTest.createInvalidNumericalLedgerIndex` could be removed because it does not make sense to test the `LedgerIndex.plus` method after bad calls to `LedgerIndex.of`.

sappenin replied:
Yes, I agree -- it looks like your new test (see [here](https://github.com/XRPLF/xrpl4j/pull/98/files#diff-ba78bf512011908df564830de4f68373ea102c5f6ec9b3dea7a1febef3936615R31)) covers everything that was covered by `createInvalidNumericalLedgerIndex` -- feel free to remove it and verify that test coverage is the same?

tekbar11 commented:
Marked constructor in `LedgerIndex` as deprecated, with a javadoc comment explaining the reason of the deprecation.

Added unit tests to `LedgerIndexTest` for `LedgerIndex`'s factory methods.

nkramer44 reviewed (approved):
Looks good to me! Thanks @tekbar11 for doing this!

nkramer44 commented on xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java:
```diff
@@ -45,8 +64,13 @@ public static LedgerIndex of(String value) {
+   * @throws NullPointerException if value is null
    */
   public static LedgerIndex of(UnsignedLong value) {
+    if (value == null) {
```
You could replace this with `Object.requireNonNull(value)`, which may make Codecov happy

sappenin replied:
Yes, also see comment above. As a style, this should be: `Objects.requireNonNull(value);`

tekbar11 replied:
done

sappenin commented on xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java:
```diff
@@ -34,9 +40,22 @@ public LedgerIndex(String value) {
-    return new LedgerIndex(value);
+  public static LedgerIndex of(String value)
+  throws NumberFormatException {
+    if (value == null) {
```
As a style, this should be: `Objects.requireNonNull(value);`

tekbar11 replied:
done

sappenin commented on xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java:
```diff
@@ -85,6 +109,24 @@ public LedgerIndex plus(LedgerIndex other) {
+      unsignedLongValue();
+      return true;
+    } catch (NumberFormatException e) {
+      return false;
```
Is this a phantom comment, or is this actually not covered? Seems like line 31 of the unit test should trigger this. Am I missing something? Might be worth "run with coverage" in Intellij just to double-check.

nkramer44 replied:
I was confused about this as well. I can pull this down and run with coverage. Sometimes Codecov isn't 100% accurate

tekbar11 replied:
Yes this should have been considered covered, but jacoco seems to not report `throw` statements as covered.
From [the jacoco FAQ](https://www.eclemma.org/jacoco/trunk/doc/faq.html):

**Source code lines with exceptions show no coverage. Why?**

*JaCoCo determines code execution with so called probes. Probes are inserted into the control flow at certain positions. Code is considered as executed when a subsequent probe has been executed. In case of exceptions such a sequence of instructions is aborted somewhere in the middle and the corresponding line of source code is not marked as covered.*

I fixed this in my latest commit, now the code coverage should be OK.

sappenin reviewed (changes requested):
This is really close - some very minor changes and I think we're good to go. Thanks for doing this!

tekbar11 commented:
I fixed the code coverage loss I introduced, jacoco now only highlights the code that was already not considered as covered before I made changes to `LexerIndex`.
